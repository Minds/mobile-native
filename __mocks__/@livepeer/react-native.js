// mock @livepeer/react-native
export const LivepeerConfig = {
  setConfig: jest.fn(),
  getConfig: jest.fn(),
  get: jest.fn(),
  set: jest.fn(),
};

export const createReactClient = jest.fn();
export const createLivepeerClient = jest.fn();
export const Player = jest.fn();
