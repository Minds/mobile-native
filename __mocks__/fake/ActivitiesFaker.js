function load(count, container) {
  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid: code,
      custom_data: false,
      custom_type: false,
      rowKey: 'something' + i,
      description: 'Congratulations! ',
      edited: '',
      guid: code,
      mature: false,
      time_created: '1522036284',
      ownerObj: {
        guid: '824853017709780997',
        type: 'user',
        subtype: false,
        time_created: '1522036284',
      },
      message: 'Message',
      title: 'TITLE',
      owner_guid: '824853017709780997',
      parent_guid: '838106762591510528',
      perma_url: false,
      thumbnail_src: false,
      type: 'activity',
      wire_totals: {
        tokens: 1000000000000000000,
      },
      containerObj: container,
    };
  });

  return {
    activities: activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa',
  };
}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load };
}

/**
 * Generate a single fake entity
 * @param {object} containerObj optional entity container
 * @returns
 */
export function fakeOne(containerObj) {
  const resp = load(1, containerObj);
  return resp.activities[0];
}
