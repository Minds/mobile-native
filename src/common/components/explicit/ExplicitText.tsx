import React, { Component } from 'react';

import { observer } from 'mobx-react';
import * as entities from 'entities';
import {
  Text,
  StyleSheet,
  View,
  Dimensions,
  Platform,
  TextStyle,
} from 'react-native';

import Tags from '../../../common/components/Tags';
import colors from '../../../styles/Colors';
import i18n from '../../services/i18n.service';
import ThemedStyles from '../../../styles/ThemedStyles';
import type ActivityModel from '../../../newsfeed/ActivityModel';
import { LinearGradient } from 'expo-linear-gradient';

type PropsType = {
  entity: ActivityModel;
  selectable?: boolean;
  navigation: any;
  style?: TextStyle | Array<TextStyle>;
  noTruncate?: boolean;
};

type StateType = {
  more: boolean;
  width: number;
  height: number;
};

@observer
export default class ExplicitText extends Component<PropsType, StateType> {
  needTruncate = false;

  state = {
    more: false,
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height,
  };

  /**
   * Handle dimensions changes
   */
  dimensionChange = () => {
    this.setState({
      width: Dimensions.get('window').width,
    });
  };

  /**
   * constructor
   */
  constructor(props) {
    super(props);
    Dimensions.addEventListener('change', this.dimensionChange);
  }

  /**
   * On component will unmount
   */
  componentWillUnmount() {
    // allways remove listeners on unmount
    Dimensions.removeEventListener('change', this.dimensionChange);
  }

  /**
   * Render
   */
  render() {
    const theme = ThemedStyles.style;
    const entity = this.props.entity;

    let title: string =
      entity.title && !entity.perma_url
        ? entities.decodeHTML(entity.title).trim()
        : '';
    let message = entity.text ? entities.decodeHTML(entity.text).trim() : '';

    if (title === message) {
      message = '';
    }

    let body: Element | null = null;
    let moreLess: JSX.Element | null = null;
    let explicitToggle = null;

    if (message !== '') {
      if (!this.props.noTruncate) {
        const truncated = this.truncate(message);
        // truncate if necessary
        if (message.length > truncated.length) {
          if (!this.state.more) message = truncated;
          moreLess = this.getMoreLess();
        }
      }

      body =
        entity.shouldBeBlured() && !entity.mature_visibility ? (
          <Text style={styles.mature}>{message}</Text>
        ) : (
          <Tags
            navigation={this.props.navigation}
            style={this.props.style}
            selectable={this.props.selectable}>
            {message}
          </Tags>
        );
    }

    const titleCmp: React.ReactNode = title ? (
      <Tags
        navigation={this.props.navigation}
        style={[
          this.props.style,
          theme.marginBottom,
          message ? styles.titleHasMessage : null,
        ]}>
        {title}
      </Tags>
    ) : null;

    return (
      <View
        style={[styles.container, !!entity.paywall ? styles.paywalled : null]}>
        {titleCmp}
        {body}
        {moreLess}
        {explicitToggle}
        {!!entity.paywall && this.renderGradient()}
      </View>
    );
  }

  renderGradient() {
    const backgroundColor = ThemedStyles.getColor('secondary_background');
    const startColor = backgroundColor + '00';
    const endColor = backgroundColor + 'FF';
    return (
      <LinearGradient colors={[startColor, endColor]} style={styles.linear} />
    );
  }

  /**
   * Truncate text
   */
  truncate(message: string) {
    const limit = this.getTextLimit();

    let truncated = false;
    let lines: Array<string> = message.split('\n');
    if (lines.length > 20) {
      lines = lines.slice(0, 20);
      truncated = true;
    }

    const strLines = lines.join('\n');
    if (strLines.length > limit) {
      return strLines.substr(0, limit) + '...';
    }
    return strLines + (truncated ? '...' : '');
  }

  /**
   * Get text char limit based on screen height
   */
  getTextLimit() {
    return this.state.height * 0.5;
  }

  /**
   * Returns more or less button
   */
  getMoreLess() {
    const msg = this.state.more ? i18n.t('showLess') : i18n.t('readMore');
    return (
      <Text style={styles.readmore} onPress={this.toggleReadMore}>
        {msg}
      </Text>
    );
  }

  /**
   * Toggle read more
   */
  toggleReadMore = () => {
    this.setState({ more: !this.state.more });
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  paywalled: {
    maxHeight: 70,
    overflow: 'hidden',
  },
  linear: {
    position: 'absolute',
    height: 50,
    width: '100%',
    left: 0,
    bottom: 0,
    zIndex: 9999,
  },
  readmore: {
    color: colors.primary,
    marginTop: 5,
  },
  mature: {
    ...Platform.select({
      ios: {
        color: 'rgba(255, 255, 255, 0.1)',
        shadowColor: 'black',
        shadowOpacity: 1,
        shadowRadius: 2,
        shadowOffset: {
          height: -1,
        },
      },
      android: {
        color: 'transparent',
        textShadowColor: 'rgba(107, 107, 107, 0.5)',
        textShadowOffset: { width: -1, height: 1 },
        textShadowRadius: 20,
      },
    }),
  },
  titleHasMessage: {
    fontWeight: 'bold',
  },
});
