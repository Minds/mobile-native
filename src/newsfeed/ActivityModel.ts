import { runInAction, action, observable, decorate } from 'mobx';
import FastImage from 'react-native-fast-image';
import { FlatList, Alert, Platform } from 'react-native';

import BaseModel from '../common/BaseModel';
import UserModel from '../channel/UserModel';
import wireService from '../wire/WireService';
import sessionService from '../common/services/session.service';
import {
  setPinPost,
  deleteItem,
  unfollow,
  follow,
} from '../newsfeed/NewsfeedService';
import api from '../common/services/api.service';

import { GOOGLE_PLAY_STORE, MINDS_CDN_URI, MINDS_URI } from '../config/Config';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import type { ThumbSize, LockType } from '../types/Common';
import type GroupModel from '../groups/GroupModel';
import { SupportTiersType } from '../wire/WireTypes';
import mindsService from '../common/services/minds.service';
import NavigationService from '../navigation/NavigationService';
import { showNotification } from '../../AppMessages';

type Thumbs = Record<ThumbSize, string> | Record<ThumbSize, string>[];

/**
 * Activity model
 */
export default class ActivityModel extends BaseModel {
  // Observable properties
  @observable pinned: boolean = false;
  @observable time_created: string = '';
  @observable message: string = '';
  @observable title: string = '';
  @observable mature: boolean = false;
  @observable edited: '0' | '1' = '0';
  @observable paywall: true | '1' | '' = '';

  // decorated observables
  'is:following': boolean;
  'thumbs:down:count': number;
  'thumbs:up:count': number;
  'comments:count': number;
  'thumbs:down:user_guids': Array<number>;
  'thumbs:up:user_guids': Array<number>;
  seen?: boolean;
  rowKey?: string;
  description?: string; // on image objects in some cases the message is on description field
  containerObj?: GroupModel;
  remind_object?: ActivityModel;
  ownerObj!: UserModel;
  listRef?: FlatList<any>;
  thumbnails?: Thumbs;
  paywall_unlocked: boolean = false;
  guid: string = '';
  subtype: string = '';
  entity_guid: string = '';
  owner_guid: string = '';
  custom_type?: string;
  custom_data?: Array<any> | any;
  nsfw?: Array<number>;
  flags?: any;
  reminds: number = 0;
  impressions: number = 0;
  perma_url?: string;
  cinemr_guid?: string;
  thumbnail_src?: string;
  dontPin?: boolean;
  boosted?: boolean;
  wire_threshold?: {
    type: 'token' | 'money';
    min: number;
    support_tier: SupportTiersType;
  } | null;
  _preview?: boolean;
  attachments?: {
    attachment_guid: string;
  };
  type?: string;
  permaweb_id?: string;
  remind_deleted?: boolean;
  remind_users?: Array<UserModel>;

  /**
   * Mature visibility flag
   */
  @observable mature_visibility: boolean = false;

  /**
   * Is visible in flat list
   */
  @observable is_visible: boolean = false;

  /**
   *  List reference setter
   */
  set _list(value) {
    this.__list = value;

    // the reminded object need access to the metadata service too
    if (this.remind_object) {
      this.remind_object._list = value;
    }
  }

  /**
   *  List reference getter
   */
  get _list() {
    return this.__list;
  }

  /**
   * Block owner
   */
  async blockOwner() {
    await super.blockOwner();
    if (this._list) {
      this._list.refresh();
    }
  }

  /**
   * Unblock owner
   */
  async unblockOwner() {
    await super.unblockOwner();
    if (this._list) {
      this._list.refresh();
    }
  }

  /**
   * Has media
   */
  hasMedia(): boolean {
    const type = this.custom_type || this.subtype;
    switch (type) {
      case 'image':
      case 'video':
      case 'batch':
        return true;
    }
    if (this.perma_url) {
      return true;
    }
    return false;
  }

  /**
   * Has an attached video
   */
  hasVideo(): boolean {
    return (this.custom_type || this.subtype) === 'video';
  }

  /**
   * Has an attached image
   */
  hasImage(): boolean {
    const type = this.custom_type || this.subtype;
    return type === 'image' || type === 'batch';
  }

  hasThumbnails(): boolean {
    return Array.isArray(this.thumbnails) && this.thumbnails.length === 0
      ? false
      : this.thumbnails
      ? true
      : false;
  }

  toPlainObject() {
    const plainEntity = super.toPlainObject();
    if (plainEntity.remind_object && plainEntity.remind_object.__list) {
      delete plainEntity.remind_object.__list;
    }

    return plainEntity;
  }

  /**
   * Child models
   */
  childModels(): any {
    return {
      ownerObj: UserModel,
      remind_object: ActivityModel,
    };
  }

  /**
   * Get the activity thumb source
   * {uri: 'http...'}
   * @param {string} size
   */
  getThumbSource(size: ThumbSize = 'medium') {
    // for gif use always the same size to take advantage of the cache (they are not resized)
    if (this.isGif()) {
      size = 'xlarge';
    }

    if (this.thumbnails && this.thumbnails[size]) {
      return { uri: this.thumbnails[size], headers: api.buildHeaders() };
    }

    // fallback to old behavior
    if (this.paywall || this.paywall_unlocked) {
      const guid = this.entity_guid === '' ? this.guid : this.entity_guid;
      const unlock = this.isOwner() ? '?unlock_paywall=1' : '';
      return {
        uri: MINDS_URI + 'fs/v1/thumbnail/' + guid + unlock,
        headers: api.buildHeaders(),
      };
    }
    if (this.custom_type === 'batch') {
      return {
        uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.entity_guid + '/' + size,
        headers: api.buildHeaders(),
      };
    }
    return {
      uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.guid + '/' + size,
      headers: api.buildHeaders(),
    };
  }

  isGif(): boolean {
    return this.custom_data && this.custom_data[0] && this.custom_data[0].gif;
  }

  /**
   * Preload thumb on image cache
   */
  preloadThumb(size: ThumbSize = 'medium') {
    FastImage.preload([this.getThumbSource(size)]);
  }

  shouldBeBlured(): boolean {
    const user = sessionService.getUser();

    if (user && user.mature) {
      return false;
    }

    if (this.nsfw !== undefined) {
      let res = [1, 2, 4].filter((nsfw) => this.nsfw!.indexOf(nsfw) > -1)
        .length;
      if (res) {
        return true;
      }
    }

    if (typeof this.flags !== 'undefined') {
      return !!this.flags.mature;
    }

    if (typeof this.mature !== 'undefined') {
      return !!this.mature;
    }

    if (
      typeof this.custom_data !== 'undefined' &&
      typeof this.custom_data[0] !== 'undefined'
    ) {
      return !!this.custom_data[0].mature;
    }

    if (this.custom_data !== undefined) {
      return !!this.custom_data.mature;
    }

    return false;
  }

  /**
   * Get activity text
   */
  get text() {
    return this.message || this.description || '';
  }

  @action
  toggleMatureVisibility = () => {
    if (GOOGLE_PLAY_STORE || Platform.OS === 'ios') {
      showNotification(i18n.t('activity.notSafeComment'));
      return;
    }
    this.mature_visibility = !this.mature_visibility;

    if (this.remind_object && this.remind_object) {
      this.remind_object.mature_visibility = this.mature_visibility;
    }
  };

  @action
  setVisible(value: boolean) {
    this.is_visible = value;
    if (this.remind_object) {
      this.remind_object.is_visible = value;
    }
  }

  /**
   * Increment the comments counter
   */
  @action
  incrementCommentsCounter() {
    this['comments:count']++;
  }

  /**
   * Decrement the comments counter
   */
  @action
  decrementCommentsCounter() {
    this['comments:count']--;
  }

  /**
   * Unlock the activity and update data on success
   */
  @action
  async unlock(ignoreError = false) {
    try {
      const result: object | false = await wireService.unlock(this.guid);
      if (result) {
        // all changes should be atomic (trigger render only once)
        runInAction(() => {
          // create a new model because we need the child models
          const list = this.__list;
          const model: ActivityModel = ActivityModel.create(result);
          Object.assign(this, model);
          this.__list = list;
        });
      }

      if (
        this.entity_guid &&
        this.perma_url &&
        this.perma_url?.startsWith(MINDS_URI)
      ) {
        NavigationService.push('BlogView', {
          guid: this.entity_guid,
          unlock: true,
        });
      }

      return result;
    } catch (err) {
      if (!ignoreError) {
        Alert.alert(err.message);
      }
      return false;
    }
  }

  /**
   * Get the lock type for the activity
   */
  getLockType = (): LockType | undefined => {
    const support_tier: SupportTiersType | null =
      this.wire_threshold && 'support_tier' in this.wire_threshold
        ? this.wire_threshold.support_tier
        : null;
    if (!support_tier) {
      return;
    }
    let type: LockType = support_tier.public ? 'members' : 'paywall';

    if (mindsService.settings.plus.support_tier_urn === support_tier.urn) {
      type = 'plus';
    }

    return type;
  };

  /**
   * Unlock the entity or prompt pay options
   */
  async unlockOrPay() {
    const result = await this.unlock(true);

    if (result) {
      return;
    }

    const lockType = this.getLockType();

    const support_tier: SupportTiersType | null =
      this.wire_threshold && 'support_tier' in this.wire_threshold
        ? this.wire_threshold.support_tier
        : null;

    switch (lockType) {
      case 'plus':
        NavigationService.push('UpgradeScreen', {
          support_tier,
          entity: this,
          onComplete: (resultComplete: any) => {
            if (resultComplete && resultComplete.payload.method === 'onchain') {
              setTimeout(() => {
                Alert.alert(
                  i18n.t('wire.weHaveReceivedYourTransaction'),
                  i18n.t('wire.pleaseTryUnlockingMessage'),
                );
              }, 400);
            } else {
              this.unlock();
            }
          },
        });
        break;
      case 'members':
      case 'paywall':
        NavigationService.push('JoinMembershipScreen', {
          entity: this,
          onComplete: (resultComplete: any) => {
            if (resultComplete && resultComplete.payload.method === 'onchain') {
              setTimeout(() => {
                Alert.alert(
                  i18n.t('wire.weHaveReceivedYourTransaction'),
                  i18n.t('wire.pleaseTryUnlockingMessage'),
                );
              }, 400);
            } else {
              this.unlock();
            }
          },
        });
    }
  }

  @action
  async togglePin() {
    // allow owners only
    if (!this.isOwner()) {
      return;
    }

    try {
      this.pinned = !this.pinned;
      await setPinPost(this.guid, this.pinned);
    } catch (e) {
      this.pinned = !this.pinned;
      Alert.alert(i18n.t('errorPinnedPost'));
    }
  }

  @action
  async deleteEntity() {
    try {
      await deleteItem(this.guid);
      this.removeFromList();
      ActivityModel.events.emit('deleteEntity', this);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  async deleteRemind() {
    try {
      await api.delete(`api/v3/newsfeed/${this.urn}`);
      this.removeFromList();
      ActivityModel.events.emit('deleteEntity', this);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  @action
  async toggleFollow() {
    const method = this['is:following'] ? unfollow : follow;
    try {
      await method(this.guid);
      runInAction(() => {
        this['is:following'] = !this['is:following'];
      });
    } catch (err) {
      logService.exception('[OffsetFeedListStore]', err);
      throw err;
    }
  }

  @action
  setEdited(value) {
    this.edited = value;
  }
}

/**
 * Define model observables
 */
decorate(ActivityModel, {
  //@ts-ignore
  'thumbs:down:count': observable,
  'thumbs:up:count': observable,
  'comments:count': observable,
  'thumbs:down:user_guids': observable,
  'thumbs:up:user_guids': observable,
});
