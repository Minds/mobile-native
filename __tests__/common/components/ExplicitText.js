import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from 'react-native';
import { shallow } from 'enzyme';
import ExplicitText from '~/common/components/explicit/ExplicitText';

import { activitiesServiceFaker } from '../../../__mocks__/fake/ActivitiesFaker';

import renderer from 'react-test-renderer';
import ActivityModel from '~/newsfeed/ActivityModel';
import sp from '~/services/serviceProvider';

jest.mock('~/services/serviceProvider');

// mock services
sp.mockService('styles');

describe('Explicit text component', () => {
  let user, comments, entity, screen;

  it('renders correctly', async () => {
    let mockResponse = activitiesServiceFaker().load(1);
    entity = ActivityModel.create(mockResponse.activities[0]);
    entity.mature = true;
    entity.mature_visibility = true;
    entity.decodeHTML = jest.fn();
    entity.shouldBeBlured = jest.fn();
    entity.toggleMatureVisibility = jest.fn();
    entity.decodeHTML.mockReturnValue('string');
    entity.shouldBeBlured.mockReturnValue(false);

    screen = shallow(<ExplicitText entity={entity} />);
    expect(screen).toMatchSnapshot();
  });
});
