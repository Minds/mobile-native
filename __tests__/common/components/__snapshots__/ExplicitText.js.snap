// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explicit text component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      null,
    ]
  }
>
  <Tags
    style={
      Array [
        undefined,
        undefined,
        null,
      ]
    }
  >
    TITLE
  </Tags>
</View>
`;
